queryable-spec.ts

```diff
@@ -0,0 +1,424 @@
+
+interface QueryableResultBindings {
+  type: 'bindings';
+  execute(opts?: { order?: QueryOperationOrder<RDF.Variable> }): Promise<Stream<Bindings>>;
+  variables: RDF.Variable[];
+  metadata(opts: { [key: string]: any }): Promise<QueryableResultMetadata<RDF.Variable>>;
+}
+    
+interface QueryableResultQuads {
+  type: 'quads';
+  execute(opts?: { order?: QueryOperationOrder<TermName> }): Promise<Stream<RDF.Quad>>;
+  metadata(opts: { [key: string]: any }): Promise<QueryableResultMetadata<TermName>>;
+}
+
+interface QueryableResultBoolean {
+  type: 'boolean';
+  execute(): Promise<boolean>;
+}
+
+interface QueryableResultVoid {
+  type: 'void';
+  execute(): Promise<void>;
+}
+
+type QueryableResult = QueryableResultBindings | QueryableResultBoolean | QueryableResultQuads | QueryableResultVoid;
+
+/*
+ * Context objects provide a way to pass additional bits information to
+ * implementors, such as but not limited to:
+ * - data sources
+ * - base IRI for IRI resolution
+ * - timestamp for expression evaluation
+ * - query language
+ * - ...
+ */
+
+// SourceType can be anything the query engine defines
+// TODO: we may consider defining some standards, like 'string', RDF.Source, ...
+interface QueryableContext<SourceType> {
+  sources: [SourceType, ...SourceType[]];
+  queryTimestamp?: Date; // Required for certain SPARQL operations such as NOW().
+  [key: string]: any;
+}
+    
+interface QueryableStringContext<SourceType> extends QueryableContext<SourceType> {
+  queryFormat?: QueryableFormat; // defaults to { language: 'SPARQL', version: '1.1', extensions: [] }
+  baseIRI?: string; // Required for parsing SPARQL queries
+};
+
+interface QueryableAlgebraContext<SourceType> extends QueryableContext<SourceType> {};
+    
+interface QueryableFormat {
+  language: string; // Like 'SPARQL'
+  version: string; // Like '1.1'
+  extensions: string[]; // TODO: leave the syntax of these extensions open for now?
+}
+
+/**
+ * Placeholder to represent SPARQL Algebra trees.
+ * Algebra typings are TBD. Reference implementations include:
+ * - https://www.npmjs.com/package/sparqlalgebrajs
+ */
+type Algebra = any;
+
+/* 
+ * Generic query interfaces. These allow engines to return any type of result
+ * object for any type of query, supporting the kind of flexibility required
+ * by engines such as Comunica.
+ */
+
+interface Queryable<SourceType, ResultType extends QueryableResult> {
+  query(query: string, context?: QueryableStringContext<SourceType>): Promise<ResultType>;
+}
+    
+interface QueryableAlgebra<SourceType, ResultType extends QueryableResult> {
+  query(query: Algebra, context?: QueryableAlgebraContext<SourceType>): Promise<ResultType>;
+}
+
+/*
+ * SPARQL-constrainted query interfaces. These interfaces guarantee that result
+ * objects are of the expected type as defined by the SPARQL spec.
+ */
+
+interface QueryableSparql<SourceType> {
+  boolean?(query: string, context?: QueryableContext<SourceType>): Promise<QueryableResultBoolean>;
```

Comment:
I'm not a big fan of optional methods. (`DataFactory.variable` being optional has proven to be quite annoying IMO)

Can we instead move the availability of these methods to a generic type expression? This would allow consumers to explicitly require the availability of certain methods in a type-safe manner.

Reply:
Big +1 on this from me, although we've previously found generics to be too complex at times. Perhaps one could simply throw a "not implemented" error?

Reply:
@gsvarovsky  ?

Reply:
Hmm. TBH I'm already feeling a bit icky about the method names being type nouns (and even JS keywords in two cases – that's a red flag for folks reading the code having to do a double-take).

Here's a suggestion: make the abstract query statement an object, and lean on `isSupported`:

```
// Actually 4 interfaces, BooleanQueryStatement, BindingsQueryStatement, etc.
// as per Term (etc.) pattern
interface QueryStatement<SourceType> {
  resultType: 'boolean' | 'bindings' | 'quads' | 'void',
  statement: string | Algebra,
  context?: QueryContext<SourceType> // Varies by result type
}

interface QueryableSparql<SourceType> {
  query(statement: BooleanQueryStatement): Promise<BooleanQuery>;
  query(statement: BindingsQueryStatement): Promise<BindingsQuery>;
  query(statement: QuadsQueryStatement): Promise<QuadsQuery>;
  query(statement: VoidQueryStatement): Promise<VoidQuery>;
}
```

Usage:
```
await engine.query({ resultType: 'boolean, statement: 'WHAT' }).isSupported()
```

It's a bit punctuation-y, but it's definitely readable IMO.

Reply:
> TBH I'm already feeling a bit icky about the method names being type nouns (and even JS keywords in two cases – that's a red flag for folks reading the code having to do a double-take).

What about renaming them to `queryBoolean`, `queryBindings`, ...?

> Here's a suggestion: make the abstract query statement an object, and lean on isSupported:
> It's a bit punctuation-y, but it's definitely readable IMO.

This would definitely work, but the downside is that it's more verbose, which is perhaps not what we want for a high-level query interface specifically made for simplifying the API?

Reply:
So, if I recall correctly the end goal of this high(er)-level interface is to give clients a way to ensure that the result type of a given query matches the type of the query itself as per the SPARQL standard. Maybe we should revisit having per-query-type methods rather than per-result-type? I was myself against this idea but it's worth keeping it in mind given we're still on the fence when it comes to this.

```
update(): Promise<VoidQuery>;
select(): Promise<BindingsQuery>;
construct(): Promise<QuadsQuery>;
```

I _think_ ensuring a specific result type given a known query type should be more manageable than ensuring a specific result type given an unknown query type. Coupled with `not implemented` / `not supported` errors, I think this might makes for a simple and straightforward solution.

Reply:
@gsvarovsky @rubensworks comments on the above? We can't lean on `isSupported()` due to https://github.com/rdfjs/query-spec/pull/7#issuecomment-988920776 .

When it comes to doing away with optional methods or not, I think we're looking at the following options:

1. keep optional methods as they are right now
2. enforce their presence but allow implementations to reject and/or return `null` / `undefined` 

When it comes to the naming, I think we're looking at the following options:

1. keep naming as it is
2. rename to `queryBindings`, `queryQuads`, ...
3. rename to `update()`, `select()`, ...

Reply:
My personal preference goes for 2. allowing implementations to reject / return `null` / `undefined` and 3. rename to align with SPARQL query types.

Reply:
> optional methods or not

If possible, I would do this via generics, so that presence of methods can be ensured in a type-safe manner.

If not possible, rejecting sounds ok to me.

> the naming

I'm ok with option 1 and 2 (no preference between the both of them).

Reply:
> I would do this via generics

Do you have an example? I can think of a few ways but they differ a lot.

Reply:
Not at hand, no 😅 
But the discussions in https://github.com/rdfjs/types/pull/24 might contain some inspiration.
Happy to work on a concrete suggestion though if you want.

Reply:
What about the following?

```ts
type SparqlQueryable<QueryType extends string | Algebra, SourceType, ResultType> = {}
  & (ResultType extends { void: true } ? {
    void(query: QueryType, context?: QueryStringContext<SourceType>): Promise<QueryVoid>;
  } : {})
  & (ResultType extends { boolean: true } ? {
    boolean: (query: QueryType, context?: QueryStringContext<SourceType>) => Promise<QueryBoolean>; 
  } : {})
  & (ResultType extends { quads: true } ? {
    quads(query: QueryType, context?: QueryStringContext<SourceType>): Promise<QueryQuads>;
  } : {})
  & (ResultType extends { bindings: true } ? {
    bindings(query: QueryType, context?: QueryStringContext<SourceType>): Promise<QueryBindings>;
  } : {})
;

const q1 = <SparqlQueryable<string, any, { bindings: true }>>{};
q1.bindings; // defined
q1.quads;    // undefined

const q2 = <SparqlQueryable<string, any, { quads: true }>>{};
q2.bindings; // undefined
q2.quads;    // defined

const q3 = <SparqlQueryable<string, any, { quads: true, bindings: true }>>{};
q3.bindings; // defined
q3.quads;    // defined
```

Reply:
Yes, I like something like that!
We could probably introduce some type aliases to do things like `SparqlQueryable<string, any, BindingsResults>`, and set a default value to enable `SparqlQueryable<string, any>`.

Reply:
Looking at making this work with generics made me even more convinced that we should rename to `select()`, `update()`; it seems to me to simply make more sense to have a `SparqlQueryable` interface declare the type of supported queries rather than the type of supported result types.

But, I think we have consensus on naming option 2., which is rename to `queryBindings()`, `queryQuads()` as that also takes care of @gsvarovsky 's point on methods being type nouns and conflicting with JS types. This is what I'll go with.